app/src/main/java/ai/elimu/appstore/synchronization/AppListAdapter.java

```diff
@@ -221,61 +235,80 @@ public void onClick(View view) {
                      */
-                    DownloadApplicationAsyncTask.ProgressUpdateCallback progressUpdateCallback =
-                            new DownloadApplicationAsyncTask.ProgressUpdateCallback() {
+                    final ProgressUpdateCallback progressUpdateCallback = new ProgressUpdateCallback() {
+                        @Override
+                        public void onProgressUpdated(final String progressText, final int progress) {
+                            downloadStatus.setDownloadProgressText(progressText);
+                            downloadStatus.setDownloadProgress(progress);
+                            appDownloadStatus.set(position, downloadStatus);
+                            uiHandler.post(new Runnable() {
                                 @Override
-                                public void onProgressUpdated(String progressText, int progress) {
-                                    downloadStatus.setDownloadProgressText(progressText);
-                                    downloadStatus.setDownloadProgress(progress);
-                                    appDownloadStatus.set(position, downloadStatus);
-                                    holder.textDownloadProgress.setText(downloadStatus
-                                            .getDownloadProgressText());
+                                public void run() {
+                                    Timber.d(progressText);
+                                    holder.textDownloadProgress.setText(progressText);
                                     holder.progressBarDownload.setProgress(progress);
                                 }
-                            };
-                    DownloadApplicationAsyncTask downloadApplicationAsyncTask = new
-                            DownloadApplicationAsyncTask(
-                            context.getApplicationContext(),
-                            holder.progressBarDownload,
-                            holder.textDownloadProgress,
-                            holder.btnInstall,
-                            holder.btnDownload,
-                            downloadCompleteCallback,
-                            progressUpdateCallback
-                    );
-                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
-                        downloadApplicationAsyncTask.executeOnExecutor(AsyncTask
-                                .THREAD_POOL_EXECUTOR, applicationVersion);
-                    } else {
-                        downloadApplicationAsyncTask.execute(applicationVersion);
-                    }
+                            });
+                        }
+                    };
+
+                    Ion.with(context).load(getFileUrl(applicationVersion))
+                            .progress(new ProgressCallback() {
+                                @Override
+                                public void onProgress(long downloaded, long total) {
```

Comment:
How often is this method called? Is the lagging occurring due to frequent executions of this method?

Reply:
This method is triggered on every downloaded byte chunk by Ion library.
The lagging occurred because of heavy computational job (calculating checksum value) on scrolling list up/down as per your previous comment. I already replaced by checking file length only and now lagging was fixed.